QEfficient/base/modeling_qeff.py

```diff
@@ -5,96 +5,307 @@
-Load any of
-These models
-by automatically
-detecting the type
-of the model]
-
-** QEFFBASEModel is abstract base class that defines the basic structure of these classes.
-** QEFFPipeline classes will stay at the same level as QEFFAutoModel in this hierarchy in future.
-"""
-
+import hashlib
+import inspect
+import json
+import logging
+import shutil
+import subprocess
+import warnings
 from abc import ABC, abstractmethod
-from enum import Enum
-from typing import Any, Dict, List, Optional
+from pathlib import Path
+from typing import Dict, List, Optional
 
-import numpy as np
+import onnx
 import torch
-from onnxruntime import InferenceSession as ORTInferenceSession
 
+from QEfficient.base.onnx_transforms import OnnxTransform
+from QEfficient.base.pytorch_transforms import PytorchTransform
 from QEfficient.generation.cloud_infer import QAICInferenceSession
+from QEfficient.utils import constants
+from QEfficient.utils.cache import QEFF_HOME, to_hashable
 
-
-# Defining placeholder ENUM for execute function
-class Runtime(Enum):
-    CPU_ORT = "CPU_ORT"
-    CPU_PT = "CPU_PT"
-    AI_100 = "AI_100"
+logger = logging.getLogger(__name__)
 
 
 class QEFFBaseModel(ABC):
     """
-    This class acts as parent class for all the varieties of model class (i.e. LLMs, SD, quantized etc.).
-    Enforces certain methods to be implemented by child classes.
+    Base class for all the model classes (i.e. LLMs, SD, quantized etc.).
+    Provides certain utility methods to be used by child classes.
 
-    All the child classes must provide way to load, transform(optimize), exoprt to ONNX etc. capabilities.
+    Class variables:
+    :_pytorch_transforms: Pytorch transformations to be applied after initialization.
+    :_onnx_transforms: ONNX transformations to be applied after ONNX export.
     """
 
+    _pytorch_transforms: List[PytorchTransform]
+    _onnx_transforms: List[OnnxTransform]
+
+    @classmethod
+    def _transform_names(cls) -> List[str]:
+        return [x.__name__ for x in cls._pytorch_transforms + cls._onnx_transforms]
+
     def __init__(self, model: torch.nn.Module) -> None:
         super().__init__()
         self.model = model
         self.onnx_path: Optional[str] = None
-        self.ort_session: Optional[ORTInferenceSession] = None
         self.qpc_path: Optional[str] = None
-        self.device_id: Optional[List[int]] = None
         self.qpc_session: Optional[QAICInferenceSession] = None
 
-    @classmethod
-    def from_pretrained(cls, pretrained_model_name_or_path: str, *args, **kwargs):
-        raise NotImplementedError("Must implement for child classes")
-
-    def run_pytorch(self, inputs):
-        raise NotImplementedError("Reached too far!!")
+        # Apply the transformations
+        any_transformed = False
+        for transform in self._pytorch_transforms:
+            self.model, transformed = transform.apply(self.model)
+            any_transformed = any_transformed or transformed
```

Comment:
Yes, that's the logic. Even if we have one transformation is applied, we're fine. That's also reflected in the variable name `any_transformed`. Anything else is expected here?

Reply:
What if any of the transform fails? It will still show transform is applied right?

Reply:
By failing, you mean throwing an exception? Exceptions will directly be raised to the user. If a transform is not applicable, it will return False. If all the transforms return False (meaning nothing was applicable), it will raise the warning.